### PR TITLE
Add option to skip entities generator and other fixes.

### PR DESCRIPTION
**Pull request by mshima**
Fixes https://github.com/jhipster/generator-jhipster/issues/13365.

- Add option to force regeneration of a entity alone.
If the entity contains any relationship, probably it will fail to regenerate correctly.
- Fixes entity with microservice
- Remove reimplementation of jdl defaults method.

---

Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [ ] Tests are added where necessary
-   [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by ecostanzi** [comment]
@mshima thank you! In addition to the new flag (which is helpful in many situations) I want to make sure we're not missing the main point of #13365:  when a entity has no relationships it should never trigger the generation of other entities. 

Taking this jdl as an example:
```
entity Blog { name String required }
entity BlogPost { title String required, content String required}
relationship OneToMany {Blog{blogPosts} to BlogPost{title}}
entity Language { name String required }
```
The actual behaviour: unless the `--skip-entities` is used all the entities are re-generated independently of the relationships. 

The **expected** behaviour is:

`jhipster entity Blog` -> regenerates `Blog` and `BlogPost`, but not `Language`
`jhipster entity BlogPost` -> regenerates `BlogPost` and `Blog`, but not `Language`
`jhipster entity Language` -> regenerates `Language` only, since Language has no relationships
`jhipster entity Blog --skip-entities` -> regenerates `Blog` only, ignores `BlogPost` because of the flag
`jhipster entity BlogPost --skip-entities` -> regenerates `BlogPost` only, ignores `Blog` because of the flag

**Comment by mshima** [comment]
@ecostanzi I understand your point but as I explained at https://github.com/jhipster/generator-jhipster/issues/13365#issuecomment-750472642:

- When preparing entity A with relationships with B, entity A injects the relationship into B's `otherRelationships`:
https://github.com/jhipster/generator-jhipster/blob/c7b5fb1196f7a7bcfae64ae699eec7507d4a1b10/generators/entity/index.js#L515-L516
That happens so entity A can have any field from entity B as `otherField` (label) correctly mapped using DTO.
- To guarantee entity B is regenerating correctly, entity A must be prepared (composed).

Looking at maps id:
- If entity A uses entity B as `@MapsId`, changing primary key from entity B should update primary key from entity A too.

Currently jdl always generates both sides of a relationship, but prompts don't.
See that no back-reference is added at:
https://github.com/jhipster/generator-jhipster/blob/c7b5fb1196f7a7bcfae64ae699eec7507d4a1b10/generators/entity/prompts.js#L1045-L1064

So at jhipster 7 back-reference is used by default.
Unless we decide that those problems should be ignored, skipping entities should not be default IMO.

**Comment by ecostanzi** [comment]
> Unless we decide that those problems should be ignored, skipping entities should not be default IMO.

@mshima I agree with you on this, but I don't understand why a C entity that has nothing to do with A or B has to be generated.

In the jdl I pasted the "C entity" is `Language` entity. Its json is:

```json
{
  "name": "Language",
  "fields": [
    {
      "fieldName": "name",
      "fieldType": "String",
      "fieldValidateRules": ["required"]
    }
  ],
  "relationships": [],
  "changelogDate": "20210120182718",
  "entityTableName": "language",
  "dto": "no",
  "pagination": "no",
  "service": "no",
  "jpaMetamodelFiltering": false,
  "fluentMethods": true,
  "readOnly": false,
  "embedded": false,
  "clientRootFolder": "",
  "applications": "*"
}
```

But when I run `jhipster entity Blog` i get this:

```
jhipster entity Blog

Found the .jhipster\Blog.json configuration file, entity can be automatically generated!
? Do you want to update the entity? This will replace the existing files for this entity, all your custom code will be overwritten Yes, re generate the entity
Found the .jhipster\BlogPost.json configuration file, entity can be automatically generated!
--> Found the .jhipster\Language.json configuration file, entity can be automatically generated!

     info Creating changelog for entities Blog,BlogPost,-->Language
identical src\main\resources\config\liquibase\changelog\20210120181805_added_entity_Blog.xml
identical src\main\resources\config\liquibase\fake-data\blog.csv
identical src\main\resources\config\liquibase\master.xml
identical src\main\resources\config\liquibase\changelog\20210120182518_added_entity_BlogPost.xml
identical src\main\resources\config\liquibase\changelog\20210120182518_added_entity_constraints_BlogPost.xml
identical src\main\resources\config\liquibase\fake-data\blog_post.csv
--> identical src\main\resources\config\liquibase\changelog\20210120182718_added_entity_Language.xml
identical src\main\resources\config\liquibase\fake-data\language.csv
    force .yo-rc.json
    force .jhipster\Blog.json
    force .jhipster\BlogPost.json
-->    force .jhipster\Language.json
identical src\main\java\com\mycompany\myapp\domain\Blog.java
identical src\main\java\com\mycompany\myapp\web\rest\BlogResource.java
identical src\main\java\com\mycompany\myapp\repository\BlogRepository.java
identical src\test\java\com\mycompany\myapp\web\rest\BlogResourceIT.java
identical src\test\java\com\mycompany\myapp\domain\BlogTest.java
identical src\main\java\com\mycompany\myapp\domain\BlogPost.java
identical src\main\java\com\mycompany\myapp\web\rest\BlogPostResource.java
identical src\main\java\com\mycompany\myapp\repository\BlogPostRepository.java
identical src\test\java\com\mycompany\myapp\web\rest\BlogPostResourceIT.java
identical src\test\java\com\mycompany\myapp\domain\BlogPostTest.java
-->  conflict src\main\java\com\mycompany\myapp\domain\Language.java
? Overwrite src\main\java\com\mycompany\myapp\domain\Language.java? (ynaxdH)
```

My doubt is, why bothering about Language entity in this case?

**Comment by mshima** [comment]
> My doubt is, why bothering about Language entity in this case?

Languages relationships are only parsed by `entity Languages` generator.
So back-references will only be populated by `entity Languages` generator and for discovering back-references before composing with `entity Languages` is tricky, specially for ids that may not be 1 step relationship.

A manual alternative is to add a `--regenerate-entities` and allow an array of other entities to be regenerated.

**Comment by ecostanzi** [comment]
> Languages relationships are only parsed by entity Languages generator.

JDL-Studio

The entity `Language` has no relationships. So it should be generated only and only if the developer runs `jhipster entity Language`. Why would a developer want to re-generate the `Language` entity when `jhipster entity Blog` is executed?

**Comment by mshima** [comment]
Because Blog has no idea if Language depends on it.
Unless someone implements back reference detection, this is the only way to guarantee correctness.

**Comment by ecostanzi** [comment]
Ok now I get it @mshima, thank you for your patience. It seems to me that a technical constraint is limiting the way the developer uses the entity generator. 

Speaking as a jhipster user. I usually regenerate more than one entity only at the very early stages of the project. After the project goes to production (or starts having more than 10 entities) I want to limit the scope of the jhipster command as much as possible. Even when using the jdl, what I do is `jhipster import-jdl --json-only` and the regenerating all the entities whose json has changed using `jhipster entity`. 

At the moment writing `jhipster entity A` has the same effect of `jhipster entity B` and the same of a potential `jhipster generate-all-entities`. I think we should take a step back and limiting the generation to a single entity until we implement the back reference detection. This is the behaviour of jhipster v6 right?

What do you think?

**Review by ecostanzi — CHANGES_REQUESTED** [review]
Keeping this open due to the discussion is going on in the PR comments.

**Comment by mshima** [comment]
> At the moment writing `jhipster entity A` has the same effect of `jhipster entity B` and the same of a potential `jhipster generate-all-entities`. I think we should take a step back and limiting the generation to a single entity until we implement the back reference detection. This is the behaviour of jhipster v6 right?

Major bumps is for breaking changes, switching to old behavior can generated broken projects, so IMO we should keep this behavior.

File: generators/entity/index.js [hunk 1]
```diff
@@ -118,6 +118,11 @@ class EntityGenerator extends BaseBlueprintGenerator {
             type: Boolean,
         });
 
+        this.option('skip-entities', {
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/entity/index.js:121]: maybe `--skip-others-entities` or `--single-entity`.

[the remark this answers is not in the document]
Comment (ecostanzi) [reply generators/entity/index.js:121]: I like `--single-entity`

**Comment by ecostanzi** [comment]
Ok then. Let's merge this after renaming the option. I'll open another ticket to implement the backreference and to discuss potential breaking changes.

**Review by ecostanzi — APPROVED** [review]

**Pull request merged**